### PR TITLE
Add option of deterministic training

### PR DESCRIPTION
**Pull request by am2145**
## Description
This PR adds the option to have deterministic training behaviour in chemprop v2. To enable this, `--pytorch-seed` now defaults to `None`. If the pytorch seed is set, then `deterministic` is set to true in `pl.trainer()`, using the code and workflow developed by @davidegraff in #755. A quick test using 'mol.csv' with the pytorch seed set to 0 and deterministic training returned exactly the same metrics to 16 decimal places, so setting deterministic does on the surface seem to resolve the issue.

## Relevant issues
#755

### CODE REVIEW

File: chemprop/cli/hpopt.py [hunk 1]
```diff
@@ -247,6 +252,7 @@ def train_model(config, args, train_loader, val_loader, logger):
         strategy=RayDDPStrategy(find_unused_parameters=True),
         callbacks=[RayTrainReportCallback()],
         plugins=[RayLightningEnvironment()],
+        deterministic=deterministic,
```
Comment (shihchengli) [inline chemprop/cli/hpopt.py:255]:
```suggestion
        deterministic=True if args.pytorch_seed else False,
```
You can remove L242-246 by changing this line of code

**Review by shihchengli — APPROVED** [review]
The changes look good. I only have a minor suggestion, which you can consider adopting or not. Thank you for submitting this PR.

File: chemprop/cli/hpopt.py (hunk 1 above)
Comment (KnathanM) [reply chemprop/cli/hpopt.py:255]: I agree but would check if PyTorch seed is not None as if it is 0 this will evaluate to False.

File: chemprop/cli/hpopt.py [hunk 2]
```diff
@@ -353,7 +359,12 @@ def main(args: Namespace):
     )
     val_loader = MolGraphDataLoader(val_dset, args.batch_size, args.num_workers, shuffle=False)
 
-    torch.manual_seed(args.pytorch_seed)
+    if args.pytorch_seed is None:
+        seed = torch.seed()
+    else:
+        seed = args.pytorch_seed
```
Comment (KnathanM) [inline chemprop/cli/hpopt.py:365]: Similar to Shih-Cheng's other comment, we could change this to
seed= PyTorch seed if PyTorch seed is not None else torch.seed()

I prefer this as it is less lines

File: chemprop/cli/hpopt.py (hunk 1 above)
Comment (am2145) [reply chemprop/cli/hpopt.py]: I have implemented this as `deterministic=True if args.pytorch_seed is not None else False` as suggested by you both.

File: chemprop/cli/hpopt.py (hunk 2 above)
Comment (am2145) [reply chemprop/cli/hpopt.py]: I have made this change.

File: chemprop/cli/hpopt.py (hunk 1 above)
Comment (davidegraff) [reply chemprop/cli/hpopt.py]: Why not
`deterministic=args.pytorch_seed is not None`

Comment (am2145) [reply chemprop/cli/hpopt.py]: Thanks for the suggestion. That's cleaner and I have made this change.

**Review by davidegraff — APPROVED** [review]
